pkg/networkserver/grpc_gsns.go

```diff
@@ -437,23 +443,30 @@ func (ns *NetworkServer) handleUplink(ctx context.Context, up *ttnpb.UplinkMessa
 	for r := bytes.NewReader(mac); r.Len() > 0; {
 		cmd := &ttnpb.MACCommand{}
 		if err := lorawan.DefaultMACCommands.ReadUplink(r, cmd); err != nil {
-			logger.
-				WithField("unmarshaled", len(cmds)).
-				WithError(err).
-				Warn("Failed to unmarshal MAC commands")
+			logger.WithFields(log.Fields(
+				"bytes_left", r.Len(),
+				"mac_count", len(cmds),
+			)).WithError(err).Warn("Failed to unmarshal MAC command")
 			break
 		}
+		logger.WithField("cid", cmd.CID).Debug("Read MAC command")
 		cmds = append(cmds, cmd)
 	}
+	logger = logger.WithField("mac_count", len(cmds))
+	ctx = log.NewContext(ctx, logger)
 
+	logger.Debug("Waiting for deduplication window to close...")
 	select {
 	case <-ctx.Done():
 		return ctx.Err()
 	case <-ns.deduplicationDone(ctx, up):
 	}
 
 	up.RxMetadata = acc.Accumulated()
+	ctx = log.NewContextWithField(ctx, "metadata_count", len(up.RxMetadata))
```

Comment:
It's cheaper to add the field to the existing logger and then update the context, instead of updating the logger in the context and then getting it from the context.